services/src/main/java/io/grpc/services/ChannelzProtoUtil.java

```diff
@@ -0,0 +1,226 @@
+import io.grpc.internal.Channelz.ServerStats;
+import io.grpc.internal.Channelz.SocketStats;
+import io.grpc.internal.Channelz.TransportStats;
+import io.grpc.internal.Instrumented;
+import io.grpc.internal.WithLogId;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * A static utility class for turning internal data structures into protos.
+ */
+final class ChannelzProtoUtil {
+  private ChannelzProtoUtil() {
+    // do not instantiate.
+  }
+
+  static ChannelRef toChannelRef(WithLogId obj) {
+    return ChannelRef
+        .newBuilder()
+        .setChannelId(obj.getLogId().getId())
+        .setName(obj.toString())
+        .build();
+  }
+
+  static SubchannelRef toSubchannelRef(WithLogId obj) {
+    return SubchannelRef
+        .newBuilder()
+        .setSubchannelId(obj.getLogId().getId())
+        .setName(obj.toString())
+        .build();
+  }
+
+  static ServerRef toServerRef(WithLogId obj) {
+    return ServerRef
+        .newBuilder()
+        .setServerId(obj.getLogId().getId())
+        .setName(obj.toString())
+        .build();
+  }
+
+  static SocketRef toSocketRef(WithLogId obj) {
+    return SocketRef
+        .newBuilder()
+        .setSocketId(obj.getLogId().getId())
+        .setName(obj.toString())
+        .build();
+  }
+
+  static Server toServer(Instrumented<ServerStats> obj)
+      throws ExecutionException, InterruptedException {
+    ServerStats stats = obj.getStats().get();
+    return Server
+        .newBuilder()
+        .setRef(toServerRef(obj))
+        .setData(toServerData(stats))
+        .build();
+  }
+
+  static ServerData toServerData(ServerStats stats) {
+    return ServerData
+        .newBuilder()
+        .setCallsStarted(stats.callsStarted)
+        .setCallsSucceeded(stats.callsSucceeded)
+        .setCallsFailed(stats.callsFailed)
+        .setLastCallStartedTimestamp(Timestamps.fromMillis(stats.lastCallStartedMillis))
+        .build();
+  }
+
+  static Socket toSocket(Instrumented<SocketStats> obj)
+      throws ExecutionException, InterruptedException {
+    SocketStats socketStats = obj.getStats().get();
+    return Socket.newBuilder()
+        .setRef(toSocketRef(obj))
+        .setRemote(toAddress(socketStats.remote))
+        .setLocal(toAddress(socketStats.local))
+        .setData(toSocketData(socketStats.data))
+        .build();
+  }
+
+  static Address toAddress(SocketAddress address) {
+    Address.Builder builder = Address.newBuilder();
+    if (address instanceof InetSocketAddress) {
+      InetSocketAddress inetAddress = (InetSocketAddress) address;
+      builder.setTcpipAddress(
+          TcpIpAddress
+              .newBuilder()
+              .setIpAddress(
+                  ByteString.copyFrom(inetAddress.getAddress().getAddress()))
+              .build());
+    } else if (address.getClass().getName().equals("io.netty.channel.unix.DomainSocketAddress")) {
```

Comment:
This should be an `endsWith` to account for shading.

Reply:
done